pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SparkRBackendHandler.scala

```diff
@@ -0,0 +1,204 @@
+            cls = Some(o.getClass)
+            obj = o
+          }
+        }
+      }
+
+      val args = readArgs(numArgs, dis)
+
+      val methods = cls.get.getMethods()
+      val selectedMethods = methods.filter(m => m.getName() == methodName)
+      if (selectedMethods.length > 0) {
+        val selectedMethod = selectedMethods.filter { x => 
+          matchMethod(numArgs, args, x.getParameterTypes())
+        }.head
+
+        val ret = selectedMethod.invoke(obj, args:_*)
+
+        // Write status bit
+        writeInt(dos, 0)
+        writeObject(dos, ret.asInstanceOf[AnyRef])
+      } else if (methodName == "<init>") {
+        // methodName should be "<init>" for constructor
+        val ctor = cls.get.getConstructors().filter { x =>
+          matchMethod(numArgs, args, x.getParameterTypes())
+        }.head
+
+        val obj = ctor.newInstance(args:_*)
+
+        writeInt(dos, 0)
+        writeObject(dos, obj.asInstanceOf[AnyRef])
+      } else {
+        throw new IllegalArgumentException("invalid method " + methodName + " for object " + objId)
+      }
+    } catch {
+      case e: Exception => {
+        System.err.println(s"$methodName on $objId failed with " + e)
+        e.printStackTrace()
+        writeInt(dos, -1)
+      }
+    }
+  }
+
+  // Read a number of arguments from the data input stream
+  def readArgs(numArgs: Int, dis: DataInputStream): Array[java.lang.Object] =  {
+    (0 until numArgs).map { arg =>
+      readObject(dis)
+    }.toArray
+  }
+
+  // Checks if the arguments passed in args matches the parameter types.
+  // NOTE: Currently we do exact match. We may add type conversions later.
+  def matchMethod(
+      numArgs: Int,
+      args: Array[java.lang.Object],
+      parameterTypes: Array[Class[_]]): Boolean = {
+    if (parameterTypes.length != numArgs) {
+      return false
+    }
+
+    for (i <- 0 to numArgs - 1) {
+      val parameterType = parameterTypes(i)
+      var parameterWrapperType = parameterType
+
+      if (parameterType.isPrimitive()) {
+        parameterWrapperType = parameterType match {
+          case java.lang.Integer.TYPE => classOf[java.lang.Integer]
+          case java.lang.Double.TYPE => classOf[java.lang.Double]
+          case java.lang.Boolean.TYPE => classOf[java.lang.Boolean]
+          case _ => parameterType
+        }
+      }
+      if(!parameterWrapperType.isInstance(args(i))) {
+        return false
+      }
+    }
+    return true
+  }
+}
+
+/**
+ * Helper singleton that tracks Java objects returned to R.
+ * This is useful for referencing these objects in RPC calls.
+ */
+object JavaObjectTracker {
+
+  // TODO: This map should be thread-safe if we want to support multiple
+  // connections at the same time
+  val objMap = new HashMap[String, Object]
+
+  // TODO: We support only one connection now, so an integer is fine.
+  // Investiage using use atomic integer in the future.
+  var objCounter: Int = 0
+
+  def getObject(id: String): Object = {
```

Comment:
So I thought about that - stylistically it didn't look correct to have an apply function on a static object (feels like a constructor). Just as an example
```
val obj = JavaObjectTracker("2")
```
looks like we are constructing a java object tracker rather than looking up things in it ?